### PR TITLE
feat: local graph view in the Dendron Side Panel

### PR DESCRIPTION
**Pull request by Harshita-mindfire**
This PR aims to add a new view in the Dendron side panel: `Graph Panel`

# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [~] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
NA
### Basics

- [~] [Write Tests](dev.process.qa.test) 
- [ ] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
NA

## Docs
- [ ] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)
 ## Analytics
Will update the airtable when PR is merged. 
- Added a GraphEvent  `GraphPanelUsed` and its properties `type`: string, `state`:string

### CODE REVIEW

**Review by kevinslin — COMMENTED** [review]
excited about this. might actually use the graph with this :)

**Comment by jonathanyeung** [comment]
Tested it out - looks very cool. 

I'm hitting an issue though - it's likely a pre-existing issue.  When I navigate a few links using the graph view, particularly in a circle (say A->B->C->B->A), sometimes the graph view ends up turning blank, and then I see some output in the webview console. Can you repro? 

```log
index.bundle.js:314916 [ERROR] 58:39.209 getLocalNoteGraphElements - Couldn't find vault of note config.ts, aborting link creation
ConsoleLogger._log @ index.bundle.js:314916
2index.bundle.js:314916 [WARN] 58:39.209 getLocalNoteGraphElements - Failed to link note vault/dendron.ref.links to vault/void. Most likely, this note does not exist.
ConsoleLogger._log @ index.bundle.js:314916
index.bundle.js:314916 [WARN] 58:39.209 getLocalNoteGraphElements - Failed to link note vault/dendron.ref.links to vault/void.*. Most likely, this note does not exist.
ConsoleLogger._log @ index.bundle.js:314916
index.bundle.js:314919 [INFO] 58:39.209 DendronNoteGraphView Object
index.bundle.js:314919 [INFO] 58:39.209 useGraphElements Object
index.bundle.js:314919 [INFO] 58:39.209 useGraphElements Object
index.bundle.js:314916 [INFO] 58:39.211 Graph - Rendering graph...
index.bundle.js:132389 You have set a custom wheel sensitivity.  This will make your app zoom unnaturally when using mainstream mice.  You should change this value from the default only if you can guarantee that all your users will use the same hardware and OS configuration as your current machine.
warn @ index.bundle.js:132389
index.bundle.js:269343 Error: Can not create edge `73eb67ea-0291-45e7-8f2f-193fd6f00643_Zq8Kz00hOPxCo4HdS5rdt` with nonexistant target `Zq8Kz00hOPxCo4HdS5rdt`
    at error (index.bundle.js:132371:9)
    at Collection.elesfn$u.restore (index.bundle.js:144566:11)
    at new Collection (index.bundle.js:144235:10)
    at Core.add (index.bundle.js:144974:18)
    at index.bundle.js:47636:25
    at Array.forEach (<anonymous>)
    at applyDisplayConfig (index.bundle.js:47634:237)
    at applyConfig (index.bundle.js:47645:646)
    at index.bundle.js:47645:955
    at fk (index.bundle.js:270846:19)
```

**Comment by Harshita-mindfire** [comment]
> I'm hitting an issue though - it's likely a pre-existing issue with graph view. When I navigate a few links using the graph view, particularly in a circle (say A->B->C->B->A), sometimes the graph view ends up turning blank, and then I see some output in the webview console. Can you repro?
> 
Hey, yes was able to repro. I have a fix for this already.

**Review by kevinslin — CHANGES_REQUESTED** [review]
1. where you able to get this to work on org-workspace? it wasn't loading for me.
1. noticed that we can toggle between local and full graph in the webview. is that on purpose? because the webview is titled "local graph". to give us more flexibility, can we just call this the `graphView` for the sidepanel?
1. i might have changed my settings but by default, it seemed to load the full graph instead of the local graph
1. can you also add some analytics around the panel?
- when its visible
- when its been clicked on

see https://github.com/dendronhq/dendron/blob/feat/local-graph-view/packages/plugin-core/src/features/TipOfTheDayWebview.ts#L63:L63

**Comment by Harshita-mindfire** [comment]
> 1. where you able to get this to work on org-workspace? it wasn't loading for me.

Yes, does load for me
image

> 2. noticed that we can toggle between local and full graph in the webview. is that on purpose? because the webview is titled "local graph". to give us more flexibility, can we just call this the `graphView` for the sidepanel?

This shouldn't be the case. The side panel should only display local graph and there should not be any filters as well as toggle option.

**Comment by hikchoi** [comment]
@kevinslin Just did a local build myself while pairing with @Harshita-mindfire. I can't repro what you are experiencing there. Might be an incomplete / old build.

File: packages/common-all/src/analytics.ts [hunk 1]
```diff
@@ -154,6 +154,7 @@ export enum AppNames {
 
 export enum GraphEvents {
   GraphThemeChanged = "GraphThemeChanged",
+  LocalGraphSidePanel = "LocalGraphSidePanel",
```
Comment (jonathanyeung) [inline packages/common-all/src/analytics.ts:157]: Can we follow the noun + verb convention?  Given that you've named it `DendronGraphPanel.tsx`, let's keep that terminology - so how about `GraphPanelUsed` for this event name? 

We can start calling this GraphPanel (side) vs GraphView (full page webview)

File: packages/plugin-core/src/views/LocalGraphView.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    Logger.info({ ctx, data: msg });
+    switch (msg.type) {
+      case GraphViewMessageEnum.onSelect: {
+        const note = this._ext.getEngine().notes[msg.data.id];
+        await new GotoNoteCommand(this._ext).execute({
+          qs: note.fname,
+          vault: note.vault,
+        });
+        AnalyticsUtils.track(GraphEvents.LocalGraphSidePanel, {
+          message: GraphViewMessageEnum.onSelect,
```
Comment (jonathanyeung) [inline packages/plugin-core/src/views/LocalGraphView.ts:65]: for consistency, let's try to use the same property names across different instances of this `GraphEvents.LocalGraphSidePanel` event type.  So maybe something like 

```typescript
AnalyticsUtils.track(GraphEvents.GraphPanelUsed, {
          type: "NodeClicked",
}
```

File: packages/plugin-core/src/views/LocalGraphView.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+              isSidePanel: true,
+            },
+            source: "vscode",
+          });
+        }
+        break;
+      }
+
+      case GraphViewMessageEnum.onLocalGraphVisible: {
+        AnalyticsUtils.track(GraphEvents.LocalGraphSidePanel, {
```
Comment (jonathanyeung) [inline packages/plugin-core/src/views/LocalGraphView.ts:108]: Rather than track 'visible' here, let's track it on the VSCode panel's visibility event.  `vscode.WebviewView.onDidChangeVisibility`

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

File: packages/common-all/src/analytics.ts (hunk 1 above)
Comment (Harshita-mindfire) [reply packages/common-all/src/analytics.ts]: updated event and changed `Local Graph View` to `Graph Panel`

**Comment by kevinslin** [comment]
> @kevinslin Just did a local build myself while pairing with @Harshita-mindfire. I can't repro what you are experiencing there. Might be an incomplete / old build.

it looks like this line that is causing the issue: https://github.com/dendronhq/dendron/blob/feat/local-graph-view/packages/plugin-core/src/views/GraphPanel.ts#L100:L100

if i don't have a custom theme or a graph theme set, `isSidePanel` is never set to `true`. instead of passing in `isSidePanel`, since we know this view is always going to be a side panel, why not create a `DendronSideGraphPanel` view that is just a wrapper around `DendronGraphPanel` with `isSidePanel` set to true?

File: packages/plugin-core/src/views/GraphPanel.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+      key: DendronTreeViewKey.GRAPH_PANEL,
+      webviewView,
+    });
+    webviewView.webview.onDidReceiveMessage(
+      this.onDidReceiveMessageHandler,
+      this
+    );
+
+    webviewView.onDidChangeVisibility(() => {
+      AnalyticsUtils.track(VSCodeEvents.GraphPanelClicked, {
```
Comment (jonathanyeung) [inline packages/plugin-core/src/views/GraphPanel.ts:56]: To reduce the number of types of events, can we use the `GraphPanelUsed` event here as well, set the type property to `VisibilityChanged` and have a state property?

Similar to what I did for Backlinks Panel: 

```typescript
    backlinkTreeView.onDidChangeVisibility((e) => {
      AnalyticsUtils.track(VSCodeEvents.BacklinksPanelUsed, {
        type: "VisibilityChanged",
        state: e.visible ? "Visible" : "Collapsed",
      });
    });
```

**Review by jonathanyeung — CHANGES_REQUESTED** [review]
One last nit on telem event naming, otherwise lgtm

**Comment by SeriousBug** [comment]
Looks good to me. I have some nitpicks, but it's better if you spin these off into a followup task so this is not blocked.

- The zoom level in the graph is not maintained between note switches. So if I zoom into the graph, then switch to another note, the zoom resets.
- The zoom level was too far by default for me, which meant note names were not visible. It looks like the zoom is tied to the size of the panel, could we cap it so at minimum it's zoomed in enough that text appears?
  


- Resizing the panel doesn't seem to resize or re-zoom the view until I switch to another note and back. It would be nice if it automatically updated as the panel is resized.

File: packages/dendron-plugin-views/src/components/DendronSideGraphPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Wrapper component around DendronGraphPanel
+ * @param props
+ * @returns DendronGraphPanel component with ide.isSidePanel set to true
+ */
+const DendronSideGraphPanel: DendronComponent = (props) => {
+  const ideDispatch = ideHooks.useIDEAppDispatch();
+  // set side panel to true
+  ideDispatch(ideSlice.actions.setIsSidePanel(true));
```
Comment (kevinslin) [inline packages/dendron-plugin-views/src/components/DendronSideGraphPanel.tsx:13]: instead of making this a state property, can we just have this as part of the `props` for `DendronGraphPanel`? this way we avoid an unnecessary dispatch

File: packages/plugin-core/src/showcase/GraphPanelTip.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+export class GraphPanelTip implements IFeatureShowcaseMessage {
+  shouldShow(_displayLocation: DisplayLocation): boolean {
+    return true;
+  }
+  get showcaseEntry(): ShowcaseEntry {
+    return ShowcaseEntry.GraphPanel;
+  }
+
+  getDisplayMessage(_displayLocation: DisplayLocation): string {
+    return "Dendron has a new Graph Panel. Check it out!";
```
Comment (jonathanyeung) [inline packages/plugin-core/src/showcase/GraphPanelTip.ts:18]: nit: Can we amend this message to maybe also help with side bar discoverability? 

"We're experimenting with a note graph panel in the new Dendron sidebar.  Check it out!"

**Review by jonathanyeung — APPROVED** [review]
